Add support for HiSilicon Hi3861 Wi-Fi SoC & HiHope Pegasus dev board

xingrz opened the pull request:
This PR brings Zephyr to HiSilicon Hi3861, a single-chip Wi-Fi SoC.

This also preliminarily brings up HiHope WiFi-IoT, one of the official boards. Currently these features are tested working:

* SysTick
* GPIO
* UART
* Pinctrl

Related PR: #57638

References:

* [HiSilicon Hi3861V100 brief](https://www.hisilicon.com/en/products/smart-iot/ShortRangeWirelessIOT/Hi3861V100)
* [HiHope Pegasus WiFi-IoT Development Suite](http://www.hihope.org/pro/pro1.aspx?mtt=8) (Chinese)
* [Hi3861V100/Hi3861LV100/Hi3881V100 WiFi SoC User Manual](https://gitee.com/HiSpark/hi3861_hdu_iot_application/raw/master/doc/manual/Hi3861V100%EF%BC%8FHi3861LV100%EF%BC%8FHi3881V100%20WiFi%E8%8A%AF%E7%89%87%20%E7%94%A8%E6%88%B7%E6%8C%87%E5%8D%97.pdf) (Chinese)

xingrz committed:
soc: riscv: hi3861: Add HiSilicon Hi3861 series

Hi3861 is a RISC-V single chip Wi-Fi SoC by HiSilicon. In the meanwhile,
Hi3861V100 is the only available model of this SoC.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
boards: riscv: Bring up HiHope WiFi-IoT dev board

HiHope Pegasus WiFi-IoT Development Suite is a set of development boards
provided by HopeRun, based on HiSilicon Hi3861V100.

This commit preliminarily brings up the board with samples/basic/blinky.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: gpio: Add GPIO driver for Hi3861

The GPIO IP on the HiSilicon Hi3861 is actually a DesignWare IP. This
commit adds a simplified implementation of the DesignWare GPIO driver
for Hi3861, with added functionality for setting pull-up and pull-down
values using the Hi3861's IOMUX.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: timer: Add timer driver for Hi3861

TIMER3 is ussed as the systick on this SoC.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
boards: hihope_wifi_iot: Enable pinctrl

The pinctrl driver takes care of the IO MUX of the chip.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: pinctrl: Add pinctrl driver for Hi3861

This commit introduces a pinctrl driver for Hi3861 that allows for
configuring the functions and settings of each pin.

Additionally, it includes a hi3861-pinctrl.h header that can be utilized
in the &pinctrl sub-node of the DTS.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
dts: riscv: hi3861: Add UART nodes

The HiSilicon Hi3861 includes three ARM PL011 UART IPs. This change adds
device tree nodes for each of these UARTs.

The existing PL011 driver has been tested and found to work well with
the Hi3861's UARTs.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
boards: hihope_wifi_iot: Enable UART console

UART0 is used as console, where the on-board CH340 USB serial adapter is
connected to.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
[DNM] manifest: Add hisilicon

The hisilicon module contains external sources imported from the
manufacturer SDK, e.g. loaderboot required by Hi3861.

**DNM: This commit will be edited once the module request is approved**

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: pinctrl: hi3861: Add private APIs for setting pull-up/pull-down

The Hi3861 uses IOMUX bits to control bias pull-up and pull-down for
GPIO pins. This commit adds private functions to the pinctrl driver that
allow the GPIO driver to set these bias values.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
dts: riscv: hi3861: Switch to own implementation of GPIO driver

So that GPIO_PULL_UP and GPIO_PULL_DOWN flags work

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
boards: hihope_wifi_iot: Add node for user button

Let the sample/basic/button works

Signed-off-by: Chen Xingyu <hi@xingrz.me>

carlocaione commented on drivers/timer/hi3861_timer.c:
```diff
@@ -0,0 +1,71 @@
+	ARG_UNUSED(arg);
+
+	// Read EOI to clear interrupt
+	uint32_t n = TIMER_3->EOI;
```
please use the helper `sys_read32` instead (same for all the other occurrences).

xingrz replied:
Are you suggesting that accessing registers using the `BASE + OFFSET` method is better than accessing struct members by casting a pointer to the base address?

carlocaione replied:
definitely. With pointer and base address you are offloading the task of creating the correct code to the compiler, using the helpers you are avoiding that. Also the helpers sometimes add some kind of optimizations, like memory barriers, etc...

carlocaione commented on soc/riscv/riscv-privilege/hi3861/soc.h:
```diff
@@ -7,3 +7,15 @@
 #include <soc_common.h>
 #include <zephyr/devicetree.h>
+
+/* Implement the following ARM intrinsics as no-op:
+ * - ARM Data Memory Synchronization Barrier
+ * - ARM Instruction Synchronization Barrier
+ */
+#ifndef __DMB
+#define __DMB()
+#endif
+
+#ifndef __ISB
+#define __ISB()
+#endif
```
why this is here.

xingrz replied:
This SoC integrated a generic PL011 UART IP from ARM. The existing driver implementation calls DMB and ISB in several places. Therefore, I need these no-op defines to allow it to compile. Should I explain the purpose above the defines here, or should I mask them with something like #ifdef CONFIG_ARCH_ARM in uart_pl011.c?

https://github.com/zephyrproject-rtos/zephyr/blob/f12ffaea1a9105d0cff74ea50ecde0549c680740/drivers/serial/uart_pl011.c#L211

carlocaione replied:
move
```c
#ifndef __DMB
#define __DMB()
#endif
```
in the driver

carlocaione reviewed (changes requested):
All the commit messages are missing.

fkokosinski commented on dts/riscv/hisilicon/hi3861v100.dtsi:
```diff
@@ -0,0 +1,112 @@
+		#clock-cells = <0>;
+	};
+
+	soc {
```
Please add SoC compat here.

fkokosinski commented on soc/riscv/riscv-privilege/hi3861/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2023 Chen Xingyu <hi@xingrz.me>
```
What's the purpose of this script? Is it to program the device? If so - this should be moved to the board CMake script.

xingrz replied:
The romcode of this SoC verifies the firmware on the flash, which must include a specific format of header (including some checksums). It is not a bare binary generated by objcopy -O binary. Also, it's not board-specific.

fkokosinski commented on soc/riscv/riscv-privilege/hi3861/pinctrl_soc.h:
```diff
@@ -0,0 +1,62 @@
+	} bit;
+} pinctrl_soc_pad_ctrl_t;
+
+typedef struct {
```
Ditto

fkokosinski commented on soc/riscv/riscv-privilege/hi3861/scripts/make_burn_bin.py:
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 2023 Chen Xingyu <hi@xingrz.me>
```
Ditto. If you need to add a custom runner, then you should add it in `scripts/west_commands/runners/`.

fkokosinski commented on west.yml:
```diff
@@ -79,6 +79,12 @@ manifest:
       path: modules/hal/gigadevice
       groups:
         - hal
+    - name: hal_hisilicon
```
Please note that you need to create an issue with a new HAL request. Otherwise this won't be pulled.

xingrz commented:
@carlocaione @fkokosinski Thanks for your review. Perhaps I should PR only the necessary Timer and GPIO changes first, and leave the more complex Pinctrl and UART stuff for the next PR?

nordicjm commented on soc/riscv/riscv-privilege/hi3861/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+set(SCRIPTS_DIR ${CMAKE_CURRENT_SOURCE_DIR}/scripts)
+set(BOOT_DIR ${CMAKE_BINARY_DIR}/modules/hal_hisilicon/hi3861/boot)
+
+if(NOT CONFIG_BOOTLOADER_MCUBOOT)
```
@tejlmand

nordicjm commented on soc/riscv/riscv-privilege/hi3861/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,28 @@
+	default 62
+
+config FLASH_LOAD_OFFSET
+	default 0xd3c0
```
Why is there a flash load offset here, does that mean if I have a bootloader at one address and an application at e.g. 0x20000, then instead of being placed at 0x20000 the application should be placed at 0x2d3c0?

xingrz replied:
The boot process of this SoC is somewhat complex. Let me explain:

1. The **romboot**, which placed at a readonly part of the chip, verifies and copies the **flashboot** to `RAM_BASE + 160KB`, then boots it up.
2. The **flashboot**, located at `FLASH_BASE + 0x00`, is built with a [few lines of open-sourced code](https://github.com/xingrz/hal_hisilicon/blob/ce6606e1f2ad5a2fbb186f8d52984d34244ced54/hi3861/boot/flashboot/startup/main.c) and some close-sourced libs. It verifies and boots the kernel at `FLASH_BASE + 0xd3c0`.

This implies some limitations:

1. Some undocumented initializations are done in the flashboot.
2. The image loaded by the romboot is not XIP-compatible, and is limited in length to `RAM_SIZE - 160KB`, which is 120KB.
3. The offset `0xd3c0` acually includes `64 bytes` of the header of the flashboot image, `52 KB` reserved for the flashboot, and `960 bytes` of metadata verified by the flashboot.

Fortunately, since the flashboot is built from source in the `hal_hisilicon` module, we can reduce the offset, which is default to `0xd3c0` in the manufacturer-provided SDK, to at least 64 KB.

nordicjm replied:
> The flashboot, located at FLASH_BASE + 0x00, is built with a [few lines of open-sourced code](https://github.com/xingrz/hal_hisilicon/blob/ce6606e1f2ad5a2fbb186f8d52984d34244ced54/hi3861/boot/flashboot/startup/main.c) and some close-sourced libs. It verifies and boots the kernel at FLASH_BASE + 0xd3c0.

This presents a problem, I don't think we can accept this for an entire soc in the tree since without this whole soc is useless, the current use of "blobs" (which requires user interaction) is to add "binary blobs" but no current in-tree board needs them to be able to boot. @carlescufi seems like a TSC discussion would be needed?

nordicjm replied:
Indeed from https://docs.zephyrproject.org/latest/contribute/bin_blobs.html
> To mitigate this risk, the scope of upstream library blobs is limited. The project maintainers define an open source test suite that an upstream target must be able to pass using only open source software included in the mainline distribution and its modules. The open source test suite currently consists of:
>
>    samples/philosophers
>
>    tests/kernel

So if this board cannot boot and run both of the above without blobs then this soc cannot be included in zephyr.

carlocaione commented:
> @carlocaione @fkokosinski Thanks for your review. Perhaps I should PR only the necessary Timer and GPIO changes first, and leave the more complex Pinctrl and UART stuff for the next PR?

I don't think that is strictly needed as long as all the commits in this PR are self contained.

xingrz committed:
soc: riscv: hi3861: Build boot images

This generates loaderboot and flashboot images, which are required for
the entire HiSilicon Hi3861 family.

The loaderboot is an executable image that are loaded to RAM by the
HiBurn runner and interacts with the runner to write the Zephyr kernel
into the flash.

The flashboot is another executable image that loaded by romboot from
flash on each boot. It's only job is to locate and boot up the Zephyr
kernel.

This commit also populates the loaderboot and flashboot paths in the
default runner arguments.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

nordicjm reviewed (dismissed):
As per https://docs.zephyrproject.org/latest/contribute/bin_blobs.html a board must be able to boot and run the 2 aforementioned samples without use of binary blobs but the description given by the submitter is that a closed source binary blob is required for this device to boot, marking as -1 until it can be clarified that this board can run the 2 samples without blobs.

xingrz commented:
> As per https://docs.zephyrproject.org/latest/contribute/bin_blobs.html a board must be able to boot and run the 2 aforementioned samples without use of binary blobs but the description given by the submitter is that a closed source binary blob is required for this device to boot, marking as -1 until it can be clarified that this board can run the 2 samples without blobs.

I think the problem should be simplified as: Is it acceptable to load Zephyr with a closed-source (or more specifically, partially closed-source) bootloader, while the Zephyr executable itself is purely built from source?

henrikbrixandersen commented:
> I think the problem should be simplified as: Is it acceptable to load Zephyr with a closed-source (or more specifically, partially closed-source) bootloader, while the Zephyr executable itself is purely built from source?

Licensing wise, there is not issue with that. The question is if we allow this in the upstream Zephyr repository. I suggest bringing this up in the Zephyr Technical Steering Committee.

xingrz commented:
Okay, the good news is that I just succeeded in booting Zephyr without using that bootloader. However, as I mentioned before, the SoC's romboot is hardcoded to copy and boot its code in RAM, which means that XIP is not available. Therefore, a bootloader (perhaps mcuboot) is still needed to point the vector back to Flash.

Nevertheless, there is hope. 🤣

fkokosinski commented on dts/riscv/hisilicon/hi3861v100.dtsi:
```diff
@@ -0,0 +1,114 @@
+	};
+
+	soc {
+		compatible = "simple-bus";
```
```suggestion
		compatible = "hisilicon,hi3861v100", "simple-bus";
```

fkokosinski commented on soc/riscv/riscv-privilege/hi3861/CMakeLists.txt:
```diff
@@ -0,0 +1,2 @@
+# Copyright (c) 2023 Chen Xingyu <hi@xingrz.me>
```
This file seems to be empty now? Remove it if it's not needed.

xingrz replied:
Yes, it's empty. But it is required by https://github.com/zephyrproject-rtos/zephyr/blob/c73133d102fb7a5d847bb2dd903e2daaffcd78a9/soc/riscv/riscv-privilege/CMakeLists.txt#L4

xingrz replied:
Maybe I can leave it an no-op `zephyr_sources()`, just like https://github.com/zephyrproject-rtos/zephyr/blob/c73133d102fb7a5d847bb2dd903e2daaffcd78a9/soc/riscv/riscv-privilege/miv/CMakeLists.txt#L3

xingrz commented:
By the way, as you can see in the DTS, this SoC integrates an ARM PL011 UART IP, which has an [existing](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_pl011.c) driver implementation in Zephyr.

I'm planning to implement the clock control feature later, but I'm not sure whether I should directly insert the platform-specific `clock_control_enable` call in `uart_pl011.c` surrounded with a pair of `#ifdef CONFIG_SOC_SERIES_HI3861 ... #endif` macro, or should I make a copy of the driver and rename it with a `_hi3861` suffix.

Can anyone please advise me on this?

xingrz committed:
scripts: runners: Add support for HiSilicon HiBurn protocol

HiBurn is a tool provided by HiSilicon for interacting with the fixed
romboot in their SoCs, such as Hi3861.

This commit adds a runner script that reimplements the HiBurn protocol
in Python, allowing for the flashing of HiSilicon SoCs using west.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

carlocaione commented on drivers/timer/hi3861_timer.c:
```diff
@@ -0,0 +1,92 @@
+
+	last_ticks++;
+
+	sys_clock_announce(1);
```
I'm still very much curious why you are going with a tickless kernel. If this is a regular counter, why not using a proper tickfull implementation?

xingrz replied:
No, it's not tickless. The timer, working at 24MHz, is set to announce `1` tick every `240000` cycles.

carlocaione replied:
you mean that you cannot set a timeout value for the timer?

xingrz replied:
It can. But I didn't implement the tickless feature. It is required?

xingrz replied:
Thank you for your suggestion. It seems that I have confused the meanings of "tickless" and "tickfull". I will update later.

carlocaione commented on drivers/serial/uart_pl011.c:
```diff
@@ -208,7 +208,9 @@ static int pl011_set_baudrate(const struct device *dev,
 	get_uart(dev)->ibrd = bauddiv >> PL011_FBRD_WIDTH;
 	get_uart(dev)->fbrd = bauddiv & ((1u << PL011_FBRD_WIDTH) - 1u);
 
+#if defined(CONFIG_ARM) || defined(CONFIG_ARM64)
 	__DMB();
+#endif
```
This should be solved when / if #57439 is merged.

xingrz committed:
boards: hihope_wifi_iot: Use hiburn as default runner

$ west flash -r hiburn --port /dev/ttyUSBx

Signed-off-by: Chen Xingyu <hi@xingrz.me>

carlocaione commented on drivers/timer/hi3861_timer.c:
```diff
@@ -0,0 +1,182 @@
+	dticks = timeout_ticks;
+	timeout_ticks = 0;
+#else
+	dticks = 1;
```
why you are announcing always 1 tick? you should announce the ticks that are passed by the last time the ISR was called. Are you sure you sure that you have not inverted the `CONFIG_TICKLESS_KERNEL` logic here?

xingrz replied:
I implemented it taking cortex_m_systick.c as reference. If I understand correctly, a "tickless" kernel expects an announcement of `N` ticks after `sys_clock_set_timeout(N)`. However, in "tickful" mode, since the timer ISR is triggered every `CYCLES_PER_TICK` cycles, an announcement for 1 tick should be made in the ISR.

carlocaione replied:
yeah, my bad. It looks like I have a bad case of Monday today.

carlocaione commented on drivers/timer/hi3861_timer.c:
```diff
@@ -0,0 +1,182 @@
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+#if defined(CONFIG_TICKLESS_KERNEL)
+	timer_stop();
```
why you need to stop the driver? You cannot stop the time form flowing right?

xingrz replied:
The systick timer on this SoC has two modes: free mode and period mode. In free mode, it operates as a one-shot timer, counting down from `0xFFFFFFFF` to zero. In period mode, it counts down from the value programmed in the `LOADCOUNT` register to zero and then repeats the process from `LOADCOUNT`. As a result, since the count cannot be programmed in free one-shot mode, I can only implement the driver using the period mode and disable it once it has been triggered.

carlocaione replied:
when operating in free mode, do you have any compare register to use? Or it just fires the interrupt when expiring to 0? In that case how you can possibly use that as timer then?

xingrz replied:
No, there is no comparing function available on this timer. This also reminds me of a new problem - I cannot determine exactly how many ticks have elapsed after an ISR is triggered and before the next timeout is set.

carlocaione replied:
@andyross @teburd do you have any other driver like this one in tree to see how this is implemented there?

xingrz replied:
Anyway, is it okay to implement just the tickful mode? Just like `leon_gptimer.c`.

carlocaione replied:
it is, especially if fully tickless is not achievable.

xingrz commented on drivers/timer/hi3861_timer.c:
```diff
@@ -0,0 +1,182 @@
+#if defined(CONFIG_TICKLESS_KERNEL)
+	timer_stop();
+
+	dticks = timeout_ticks;
```
As I mentioned in my previous reply, the timer is a countdown timer, which means that I cannot read how many cycles have elapsed, but only how many cycles are remaining. Perhaps I should pass the `timeout_ticks` as a parameter of the ISR callback? However, if I do it that way, the ISR callback would need to be connected dynamically on each call of `sys_clock_set_timeout`.

xingrz commented:
No idea why the twister keep failing at `tests/drivers/fuel_gauge/sbs_gauge`. It should be filtered out by `dt_compat_enabled("sbs,sbs-gauge-new-api")`.

xingrz commented on dts/bindings/cpu/hisilicon,riscv.yaml:
```diff
@@ -0,0 +1,8 @@
+
+description: HiSilicon RISC-V CPU
+
+compatible: "HiSilicon,riscv"
```
Fixed it. The editor's autocompletion played a trick on me.

xingrz commented on soc/riscv/riscv-privilege/hi3861/soc.c:
```diff
@@ -0,0 +1,35 @@
+static int soc_hi3861_init(void)
+{
+	/*
+	 * WDT was enabled by the romboot to check if the application is up and
+	 * running. It is expected to be disabled by default in Zephyr, unless
+	 * the WDT driver is explicitly enabled.
+	 */
+	sys_write32(0, GLB_CTL_GLB_WDT_RST_SEL_REG);
```
I think this should remain disabled, until a WDT driver is available for this SoC to make its behavior configurable.

xingrz commented on dts/riscv/hisilicon/hi3861v100.dtsi:
```diff
@@ -77,5 +77,10 @@
 			ngpios = <15>;
 		};
+
+		pinctrl: pinctrl@5000a000 {
+			compatible = "hisilicon,hi3861-pinctrl";
+			reg = <0x5000a000 0x1000>;
```
Fixed

xingrz commented on dts/riscv/hisilicon/hi3861v100.dtsi:
```diff
@@ -0,0 +1,81 @@
+				write-block-size = <16>;
+				erase-block-size = <512>;
+			};
+		};
```
Fixed

xingrz commented on dts/riscv/hisilicon/hi3861v100.dtsi:
```diff
@@ -0,0 +1,81 @@
+			reg = <0x000d8000 0x46000>;
+		};
+
+		systick: timer@40050000 {
+			compatible = "hisilicon,hi3861-timer";
+			reg = <0x40050000 0x100>;
+			clocks = <&tcxo>;
+			interrupts = <7>;
+			interrupt-names = "MACHINE_TIMER_IRQ";
+		};
```
Fixed. Thanks for your suggestions.

carlocaione commented on soc/riscv/riscv-privilege/hi3861/soc.c:
```diff
@@ -0,0 +1,24 @@
+#include <zephyr/init.h>
+#include <zephyr/arch/common/sys_io.h>
+
+#define GLB_CTL_BASE                (0x50000000)
+#define GLB_CTL_GLB_WDT_RST_SEL_REG (GLB_CTL_BASE + 0x30)
```
hardcoding an address and directly accessing that is really something that we should avoid when possible. What peripheral does this address belong to? What's GLB?

carlocaione replied:
alright, maybe you could use a syscon device for this block?

xingrz replied:
Great idea, I will update it later. Thanks!

carlocaione reviewed (changes requested):
LGTM. I'm blocking this just to see if we manage to have #57439 merged.

xingrz commented:
Updated. As discussed in #57638, flashboot sources had been moved into the `soc/` directory of this SoC.

xingrz commented:
@carlocaione By the way, as you can see in the `GLB_CTL` change above, I put the definitions of the registers under the SoC's include directory. Is it a good idea to also define the offsets of other peripherals, such as the `timer` and `pinctrl` of this SoC, under the include directory instead of inside the driver source as well?

xingrz commented:
> Actually I would entirely remove `glb_ctl.h` given that you are only using one single offset and just move the offset in `soc.c`.

But what if some of these offsets are needed for other purposes in the future? For example, the Flash driver might need to set up its clock source using `GLB_CTL`, `PMU_CMU_CTL`, and several other registers, or enable certain internal regulators with `CLDO_CTL`, etc. Should I define only the offsets that are used in the specific locations where they are needed?

carlocaione commented on soc/riscv/riscv-privilege/hi3861/soc.c:
```diff
@@ -0,0 +1,64 @@
+
+#define GLB_CTL_SYS_CTL_ID_OFFSET             (0x0)
+#define GLB_CTL_AON_SOFT_RST_W_OFFSET         (0x20)
+#define GLB_CTL_SOFT_RST_WCPU_OFFSET          (0x24)
+#define GLB_CTL_SOFT_GLB_RST_OFFSET           (0x28)
+#define GLB_CTL_GLB_WDT_RST_SEL_OFFSET        (0x30)
+#define GLB_CTL_WDT_RST_SEL_OFFSET            (0x34)
+#define GLB_CTL_AON_CKEN_OFFSET               (0x40)
+#define GLB_CTL_GLB_AON_32K_CLKEN_OFFSET      (0x50)
+#define GLB_CTL_A32K_DIV_OFFSET               (0x70)
+#define GLB_CTL_TCXO_DIV_OFFSET               (0x74)
+#define GLB_CTL_AON_PERP_CLKSEL_W_OFFSET      (0x90)
+#define GLB_CTL_RC_32K_TCXO_SEL_OFFSET        (0x94)
+#define GLB_CTL_AON_32K_SEL_OFFSET            (0x98)
+#define GLB_CTL_SYS_TICK_CFG_OFFSET           (0xC0)
+#define GLB_CTL_SYS_TICK_VALUE_0_OFFSET       (0xD0)
+#define GLB_CTL_SYS_TICK_VALUE_1_OFFSET       (0xD4)
+#define GLB_CTL_SYS_TICK_VALUE_2_OFFSET       (0xD8)
+#define GLB_CTL_SYS_TICK_VALUE_3_OFFSET       (0xDC)
+#define GLB_CTL_CLKMUX_STS_OFFSET             (0x110)
+#define GLB_CTL_DEBUG_CLKEN_OFFSET            (0x170)
+#define GLB_CTL_SOFT_INT_EN_OFFSET            (0x280)
+#define GLB_CTL_SOFT_INT_SET_OFFSET           (0x284)
+#define GLB_CTL_SOFT_INT_CLR_OFFSET           (0x288)
+#define GLB_CTL_SOFT_INT_STS_OFFSET           (0x28C)
+#define GLB_CTL_INT_SEL_OFFSET                (0x290)
+#define GLB_CTL_REFCLK_FEQ_STATUS_OFFSET      (0x358)
+#define GLB_CTL_EXT_TSF_CTRL_OFFSET           (0x400)
+#define GLB_CTL_CALI_32K_TCXO_CTL_OFFSET      (0x800)
+#define GLB_CTL_CALI_32K_TCXO_CNT_L_OFFSET    (0x810)
+#define GLB_CTL_CALI_32K_TCXO_CNT_H_OFFSET    (0x814)
+#define GLB_CTL_CALI_32K_TCXO_RESULT_L_OFFSET (0x818)
+#define GLB_CTL_CALI_32K_TCXO_RESULT_H_OFFSET (0x81C)
+#define GLB_CTL_AON_ICM_PRIORITY_OFFSET       (0xF30)
+#define GLB_CTL_MEM_CLK_FORCE_ON_OFFSET       (0xF50)
+#define GLB_CTL_MARGIN_ADJ_OFFSET             (0xF54)
+#define GLB_CTL_MARGIN_ADJ_AB_OFFSET          (0xF58)
+#define GLB_CTL_DEFAULT_SLV_EN_OFFSET         (0xF84)
+#define GLB_CTL_DEFAULT_SLV_HIT_STATUS_OFFSET (0xF88)
+#define GLB_CTL_DEFAULT_SLV_HIT_CLR_OFFSET    (0xF8C)
```
Can you leave only the offsets actually used?

nordicjm commented:
@xingrz is this board now usable for building and flashing zephyr applications without usage of any blobs and non-apache 2 code?

xingrz commented:
> 

Yes.

carlocaione reviewed (commented):
LGTM but blocked on the HAL

xingrz commented:
Rebased onto the latest `main` to use the barriers API.

fkokosinski commented:
@xingrz could you rebase and fix compliance issues? Thanks!

xingrz commented:
```
Error: See https://docs.zephyrproject.org/latest/guides/kconfig/index.html for more details.

 Undefined Kconfig symbols:

 warning: undefined symbol SOC_HI3861V100:

- Referenced at boards/riscv/hihope_wifi_iot/Kconfig.board:4:

config BOARD_HIHOPE_WIFI_IOT
	bool "HiHope WiFi-IoT"
	depends on SOC_HI3861V100 && <choice>
 Found references to undefined Kconfig symbols. If any of these are false
positives, then add them to UNDEF_KCONFIG_WHITELIST in /home/runner/work/zephyr/zephyr/./scripts/ci/check_compliance.py.

If the reference is for a comment like /* CONFIG_FOO_* */ (or
/* CONFIG_FOO_*_... */), then please use exactly that form (with the '*'). The
CI check knows not to flag it.

More generally, a reference followed by $, @, {, *, or ## will never be
flagged.

CONFIG_SOC_HI[38](https://github.com/zephyrproject-rtos/zephyr/actions/runs/5421608526/jobs/9857256386?pr=57363#step:9:39)61V100                      boards/riscv/hihope_wifi_iot/hihope_wifi_iot_defconfig:5
CONFIG_SOC_SERIES_HI3861                   boards/riscv/hihope_wifi_iot/hihope_wifi_iot_defconfig:4
```

```
Error: See https://www.pylint.org/ for more details

 E9901:Argument parser with abbreviations is disallowed (argument-parser-with-abbreviations)
File:soc/riscv/riscv-privilege/hi3861/sign_bin.py
Line:14
Column:13
```

@fkokosinski It's weird that `SOC_HI3861V100` is already defined in `hi3861/Kconfig.soc` and that the `hi3861/sign_bin.py` didn't use abbreviation argument at all.

henrikbrixandersen commented on drivers/gpio/gpio_hi3861.c:
```diff
@@ -0,0 +1,231 @@
+#include <zephyr/arch/common/sys_io.h>
+#include <zephyr/irq.h>
+
+#define GPIO_BASE DT_INST_REG_ADDR(0)
```
Will this IP always be single-instance? If not, the driver needs to support multiple instances.

xingrz replied:
In fact it's a simplified re-implementation of `gpio_dw.c`. It's Hi3861-specific because it uses the SoC's pinctrl driver to set up internal pull-up/pull-down. So, at least in Hi3861, it's single instance.
